scipy/spatial/transform/_rotation_xp.py

```diff
@@ -0,0 +1,1138 @@
+    # First compute cross products between lv and pv
+    lv_expanded = lv[..., :, None, :]
+    pv_expanded = pv[..., None, :, :]
+    cross_lp = xp.linalg.cross(lv_expanded, pv_expanded)
+    # Then compute dot product with rv
+    term5 = xp.sum(cross_lp[..., :, :, None, :] * rv[..., None, None, :, :], axis=-1)
+    # Combine all terms with proper shape alignment
+    qs = xp.abs(term1 - term2 - term3 - term4 - term5)
+    qs = xp.reshape(xp.moveaxis(qs, 1, 0), (qs.shape[1], -1))
+
+    # Find best indices from scalar components
+    max_ind = xp.argmax(xp.reshape(qs, (qs.shape[0], -1)), axis=1)
+    left_best = max_ind // rv.shape[0]
+    right_best = max_ind % rv.shape[0]
+    # Array API limitation: Integer index arrays are only allowed with integer indices
+    # TODO: Can we somehow avoid this?
+    all_idx = xp.reshape(xp.arange(left.shape[-1]), (1, -1))
+    left_idx = xp.reshape(left_best, (-1, 1))
+    left = left[left_idx, all_idx]
+    right_idx = xp.reshape(right_best, (-1, 1))
+    right = right[right_idx, all_idx]
+
+    # Reduce the rotation using the best indices
+    reduced = compose_quat(left, compose_quat(p, right))
+
+    if left is None:
+        left_best = None
+    if right is None:
+        right_best = None
+    return reduced, left_best, right_best
+
+
+def apply(quat: Array, points: Array, inverse: bool = False) -> Array:
+    xp = array_namespace(quat)
+    mat = as_matrix(quat)
+    inverse = xp.asarray(inverse, device=device(quat))
+    # We do not have access to einsum. To avoid broadcasting issues, we add a singleton
+    # dimension to the points array and remove it after the operation.
+    # TODO: We currently evaluate both branches of the where statement. For eager
+    # execution models, this may significantly slow down the function. We should check
+    # that compilers can optimize the where statement (e.g. in jax) and check if we can
+    # have an eager version that only evaluates the branch that is needed.
+    points = points[..., None]
+    if not broadcastable(mat.shape, points.shape):
+        raise ValueError(
+            "Expected equal numbers of rotations and vectors "
+            ", or a single rotation, or a single vector, got "
+            f"{mat.shape[0]} rotations and {points.shape[0]} vectors."
+        )
+
+    return xp.where(inverse, mat.mT @ points, mat @ points)[..., 0]
+
+
+def setitem(
+    quat: Array, value: Array, indexer: int | slice | EllipsisType | None
+) -> Array:
+    quat = xpx.at(quat)[indexer, ...].set(value)
+    return quat
+
+
+def align_vectors(
+    a: Array, b: Array, weights: Array | None = None, return_sensitivity: bool = False
+) -> tuple[Array, Array, Array]:
+    xp = array_namespace(a)
+    # Check input vectors
+    dtype = xp_result_type(a, force_floating=True, xp=xp)
+    a_original = xp.asarray(a, dtype=dtype, copy=True)
+    b_original = xp.asarray(b, dtype=dtype, copy=True)
```

Comment:
```suggestion
    dtype = xp_result_type(a, b, force_floating=True, xp=xp)
    a_original = xp.asarray(a, dtype=dtype)
    b_original = xp.asarray(b, dtype=dtype)
```
Please test float32 vs float64

Reply:
Do you mean we should test what happens when this is applied to `a` with dtype `float32` and `b` with dtype `float64`? The `float32` tests pass.

Reply:
correct. Upstream of my suggestion, if a is float32 and b is float64, the result dtype is float32 which isn't right.